src/relations/osm_mysql.py

```diff
@@ -0,0 +1,153 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Library containing the implementation of the legacy osm-mysql relation."""
+
+import json
+import logging
+
+from ops.charm import CharmBase, RelationBrokenEvent, RelationCreatedEvent
+from ops.framework import Object
+from ops.model import BlockedStatus
+
+from constants import LEGACY_OSM_MYSQL, PASSWORD_LENGTH
+from utils import generate_random_password
+
+logger = logging.getLogger(__name__)
+
+
+class MySQLOSMRelation(Object):
+    """Encapsulation of the legacy osm-mysql relation."""
+
+    def __init__(self, charm: CharmBase):
+        super().__init__(charm, LEGACY_OSM_MYSQL)
+
+        self.charm = charm
+
+        self.framework.observe(self.charm.on.leader_elected, self._on_leader_elected)
+        self.framework.observe(
+            self.charm.on[LEGACY_OSM_MYSQL].relation_created, self._on_osm_mysql_relation_created
+        )
+        self.framework.observe(
+            self.charm.on[LEGACY_OSM_MYSQL].relation_broken, self._on_osm_mysql_relation_broken
+        )
+
+    def _get_or_set_password_in_peer_databag(self, username: str) -> str:
+        """Get a user's password from the peer databag if it exists, else populate a password.
+
+        Args:
+            username: The mysql username
+
+        Returns:
+            a string representing the password for the mysql user
+        """
+        peer_databag = self.charm._peers.data[self.charm.app]
+
+        if peer_databag.get(f"{username}_password"):
+            return peer_databag.get(f"{username}_password")
+
+        password = generate_random_password(PASSWORD_LENGTH)
+        peer_databag[f"{username}_password"] = password
+
+        return password
+
+    def _on_leader_elected(self, _) -> None:
+        """Handle the leader elected event.
+
+        Retrieves relation data from the peer relation databag and copies
+        the relation data into the new leader unit's databag.
```

Comment:
Question: why store everything in the leaders databag rather than in the app data bag? Then you might not need t o copy things over all the time?

Reply:
The issue is that reactive charms (which the osm charms are) read relation data from the leader unit's databag. The `leader-elected` event handler exists to make sure that if a new leader is elected, the relation data is copied from the app databag to the new leader's databag